### PR TITLE
feat: mcp tool tracing

### LINKED ISSUE #10887: [FEATURE] MCP tool tracing
### Problem Statement

Span type `MCP_TOOL_CALL` is defined but never used during actual MCP tool execution.

### Proposed Solution

* if you start with this test MCP setup: https://mastra.ai/docs/mcp/overview#using-mcpclient-with-an-agent
* and use this agent setup: https://mastra.ai/docs/mcp/overview#using-mcpclient-with-an-agent

when the agent calls one of these tools, the tool call span is a `MCP_TOOL_CALL` span instead of a `TOOL_CALL` span.  

The `MCP_TOOL_CALL` includes additional details about the MCP connection that the normal span does not.

### Component

Other

### Alternatives Considered

_No response_

### Example Use Case

As a developer, I want to use MCP tools and see them in my traces.

### Additional Context

_No response_

### Verification

- [x] I have searched the existing issues to make sure this is not a duplicate
- [x] I have provided sufficient context for the team to understand the request

### PR DESCRIPTION
**Pull request by intojhanurag**
## Description

MCP tool calls now produce **`MCP_TOOL_CALL`** spans instead of generic `TOOL_CALL` spans in traces. When an agent invokes a tool from an MCP server, the trace automatically distinguishes it with a dedicated span type, MCP server name, and server version — no user code changes required.

### What changed

- **`@mastra/mcp`** — MCP client attaches `mcpMetadata` (`serverName`, `serverVersion`) to every tool it
creates
- **`@mastra/core`** — `CoreToolBuilder` detects `mcpMetadata` and creates an `MCP_TOOL_CALL` span
(replacing the `TOOL_CALL` span) with MCP-specific attributes
- **`@mastra/playground-ui`** — Trace timeline renders a distinct MCP icon and color for `mcp_tool_call`
spans

### How it works

Agent span
└─ MCP_TOOL_CALL span          ← was TOOL_CALL
 ├─ type: mcp_tool_call
 ├─ name: mcp_tool: 'list_files' on 'filesystem'
 ├─ mcpServer: filesystem
 └─ serverVersion: 1.0.0

Metadata flows automatically: `MCPClient.tools()` → `mcpMetadata` on tool object → `CoreToolBuilder` reads
it at span creation time. Users don't pass anything extra.

Screenshot 2026-02-19 220325

image

## Related Issue(s)
Fixes #10887

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [x] Test update

## Checklist

- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works

## Summary by CodeRabbit

* **New Features**
  * Tool calls from MCP servers now generate a distinct MCP span type showing server name and optional version in traces; UI displays an MCP-specific icon and styling.

* **Documentation**
  * Tracing docs updated to describe MCP_TOOL_CALL span behavior and UI representation.

* **Tests**
  * Added tests validating MCP metadata propagation and MCP span generation.

* **Chores**
  * Minor dependency bumps and test/style cleanups.

### CODE REVIEW

**Review by epinzur — CHANGES_REQUESTED** [review]
Nice feature — the metadata flow from `MCPClient` → `Tool` → `CoreToolBuilder` → span is clean, and the existing `MCP_TOOL_CALL` / `MCPToolCallAttributes` infrastructure was already in place so this plugs in naturally. Tests cover good edge cases (missing serverVersion, Vercel tool exclusion). A few things to address:

---

### 1. Repeated `(tool as any).mcpMetadata` casts in `builder.ts`

`mcpMetadata` is accessed via `(tool as any)` three times. This is fragile — extract the metadata once with a proper type instead:

```ts
// Before
const isMcpTool = !isVercelTool(tool) && 'mcpMetadata' in tool && !!tool.mcpMetadata;
// ...then later:
(tool as any).mcpMetadata.serverName  // repeated 3x

// After
const mcpMeta =
  !isVercelTool(tool) && 'mcpMetadata' in tool
    ? (tool as { mcpMetadata?: { serverName: string; serverVersion?: string } }).mcpMetadata
    : undefined;
// ...then use mcpMeta.serverName, mcpMeta.serverVersion directly
```

This avoids the repeated unsafe casts and gives you type narrowing for free.

---

### 2. MCP spans lose `toolDescription` and `toolType` attributes

When `isMcpTool` is true, the span attributes only contain `mcpServer` and `serverVersion`, dropping `toolDescription` and `toolType`. The `MCPToolCallAttributes` interface doesn't require them, but having the tool description is useful for debugging. Worth considering whether to include it.

---

### 3. Unnecessary `as any` casts in MCP client tests

In `packages/mcp/src/client/client.test.ts`, the tests use `(greetTool as any).mcpMetadata`. Since `mcpMetadata` is now part of the `Tool` type (and `client.tools()` returns `Record<string, Tool<...>>`), these casts aren't needed:

```ts
// Before
expect((greetTool as any).mcpMetadata).toBeDefined();
expect((greetTool as any).mcpMetadata.serverName).toBe('metadata-test-client');

// After
expect(greetTool.mcpMetadata).toBeDefined();
expect(greetTool.mcpMetadata!.serverName).toBe('metadata-test-client');
```

---

### 4. Changeset structure

The changeset bundles all three packages (`@mastra/core`, `@mastra/mcp`, `@mastra/playground-ui`) into one file. Since the changes are meaningfully different per package (core adds detection logic, mcp injects metadata, playground-ui adds icons), splitting into separate per-package changesets would produce cleaner changelogs.

---

### Minor nits (non-blocking)

- The `'mcp'` entry placement before `'tool'` in `spanTypePrefixes` is correct — `mcp_tool_call` would otherwise match `tool` first via the `split('_')[0]` logic. Good ordering.
- The dead code cleanup in `shared.tsx` (removing unreachable `return null` after an if/else) is a nice bugfix.

---

**Overall**: Approve with the cast cleanup and changeset split. The rest is optional.

File: packages/core/src/tools/tool-builder/builder.ts [hunk 1]
```diff
@@ -321,19 +321,29 @@ export class CoreToolBuilder extends MastraBase {
...
-          toolDescription: options.description,
-          toolType: logType || 'tool',
-        },
+        attributes: isMcpTool
+          ? {
+              mcpServer: (tool as any).mcpMetadata.serverName,
+              serverVersion: (tool as any).mcpMetadata.serverVersion,
+            }
+          : {
+              toolDescription: options.description,
```
Comment (epinzur) [inline packages/core/src/tools/tool-builder/builder.ts:359]: if we have `description` for the mcpTools, lets keep it. Feel free add new properties to the `MCPToolCallAttributes` if you think it makes sense to include them.

File: packages/core/src/tools/types.ts [hunk 2]
```diff
@@ -374,6 +374,15 @@ export interface ToolAction<
...
   providerOptions?: Record<string, Record<string, unknown>>;
+  /**
+   * Metadata identifying this tool as originating from an MCP server.
+   * Set automatically by the MCP client when creating tools.
+   * Used by CoreToolBuilder to create MCP_TOOL_CALL spans instead of TOOL_CALL spans.
+   */
+  mcpMetadata?: {
+    serverName: string;
+    serverVersion?: string;
+  };
```
Comment (epinzur) [inline packages/core/src/tools/types.ts]: can you create a type for this that is shared in both places?

**Review by epinzur — APPROVED** [review]
